SubT - allow teambase to kill all robots

m3d opened the pull request:
Yet another SubT (dangerous) PR ... we could use TEAMBASE to terminate all robots, i.e. when teambase ends everything ends ... I am not sure if it is safe for the competition (the answer is NO), but for testing on AWS it can speedup cycle time (and make logs reasonable).

Note, "minor" detail, that `teambase.py` does not parse "robot_name", because it is not receiving "origin" ... ideas welcome ...

jisa commented:
While team base is becoming integrated, I have a question: How are we going to run a base station plus a robot locally?

The reason I am asking is that last time I asked how to run multiple robots locally, it turned out to be rather complicated. Is this going to be the same story?

m3d commented:
> While team base is becoming integrated, I have a question: How are we going to run a base station plus a robot locally?
> 
> The reason I am asking is that last time I asked how to run multiple robots locally, it turned out to be rather complicated. Is this going to be the same story?

well, good point - I did not try it for a long time and I am not sure if separate docker networks are needed. I would try two bridges -> 5 terminals in total ... not great, but teambase should be almost dummy

zwn commented:
> Note, "minor" detail, that `teambase.py` does not parse "robot_name", because it is not receiving "origin" ... ideas welcome ...

It can just as easily get the robot name as an argument. Oh, it has no arguments - then it needs to be somehow wrestled into some config json as init data.

zwn commented:
A bit more general question would be - under which conditions teambase quits? In my devel branch I had a setup for a single robot testing - when at the start robot detects using robot-to-robot comms that it is solo run, it calls quits at the end. This allowed me to do a quick turnaround cloudsim tests. It seemed a bit fragile since it is unknown when all robots are loaded so I didn't submit it.

However I had an idea how to improve it. Currently the robot names follow pattern `<name>F(<time>)+(L|R)`. We could change it to `<count><name>F(<time>)+(L|R)` where `<count>` would be the total number of robots in the group. Then if the robot name starts with 1, the robot would quit the simulation by itself. If the number is bigger, it would wait in the start area until it sees the said number of robots. Any one of them then could quit the simulation (they might even coordinate using the comms). For this solution no teambase is needed -> simpler for local development. I couldn't then come up with an algorithm that the teambase could implement but which couldn't be implemented by the robots.

zwn commented:
> > While team base is becoming integrated, I have a question: How are we going to run a base station plus a robot locally?
> > The reason I am asking is that last time I asked how to run multiple robots locally, it turned out to be rather complicated. Is this going to be the same story?
> 
> well, good point - I did not try it for a long time and I am not sure if separate docker networks are needed. I would try two bridges -> 5 terminals in total ... not great, but teambase should be almost dummy

https://github.com/osrf/subt/wiki/Cloudsim%20Docker%20Compose
https://github.com/osrf/subt/blob/master/docker/docker-compose.yml
https://github.com/osrf/subt/blob/master/docker/json2docker.rb

So, it should be possible but for sure it does not look easy. I have slight preference for _there are bigger fish to fry_ (=other more pressing problems in a need of solving).

m3d commented:
testing image `200670743174.dkr.ecr.us-east-1.amazonaws.com/subt/robotika:ver69tb`

m3d commented:
Note, that locally it called the service `/subt/finish` but with result false :(

m3d commented:
Result for `T200`:
```
{
  "sim_time_duration_avg": 204,
  "sim_time_duration_std_dev": 0,
  "real_time_duration_avg": 1425,
  "real_time_duration_std_dev": 0,
  "model_count_avg": 3,
  "model_count_std_dev": 0
}
```

m3d commented:
Note, that this is important feature for my testing as I have limit  of 1GB (or it seems so) download per file .... and now I can shorten the results and do faster review. In real competition I would use T3600 or simply disable calling `/subt/finish` as we did last time ...

zwn commented on subt/teambase.py:
```diff
@@ -9,9 +9,19 @@ def __init__(self, config, bus):
+        self.start_time = None  # unknown
+        self.finish_time = None  # infinite
+        if self.robot_name is not None:
+            self.finish_time = int(self.robot_name[1:].split('F')[0])  # T100 is accepted for example
```
This seems unexpected. Why not the simpler
```suggestion
            self.finish_time = int(self.robot_name[1:])  # T100 is accepted for example
```

zwn commented on subt/teambase.py:
```diff
@@ -20,4 +30,27 @@ def update(self):
+    cfg = config_load(*args.config)
+    cfg['robot']['modules']['app']['init']['robot_name'] = args.robot_name
+    prefix = os.path.basename(args.config[0]).split('.')[0] + '-'
+    record(cfg, prefix, args.log)
```
Is the log being uploaded to the bridge?

m3d replied:
yes, that was the reason to imitate what normal robots do, i.e. almost reuse the command line arguments

zwn reviewed (approved):
Overall I think it's looking good (see the comments for details).

zwn commented:
> Note, that this is important feature for my testing as I have limit of 1GB (or it seems so) download per file .... and now I can shorten the results and do faster review. In real competition I would use T3600 or simply disable calling `/subt/finish` as we did last time ...

T3600 would not be correct since the number is absolute time and not relative from when the competition round started. For that we would need to explore https://github.com/osrf/subt/wiki/api#competition-clock

m3d commented:
> > Note, that this is important feature for my testing as I have limit of 1GB (or it seems so) download per file .... and now I can shorten the results and do faster review. In real competition I would use T3600 or simply disable calling `/subt/finish` as we did last time ...
> 
> T3600 would not be correct since the number is absolute time and not relative from when the competition round started. For that we would need to explore https://github.com/osrf/subt/wiki/api#competition-clock

Well here I changed absolute clock to relative and introduced `self.start_time`  ... i.e. I wanted to record 10s, but I could not ... so here the time is relative since Teambase start (first received `sim_time_sec`). You are right, that "just to be sure" we should use `T9999` or more.
